arctic/async/async_utils.py

```diff
@@ -0,0 +1,64 @@
+import time
+import uuid
+from enum import Enum
+
+from arctic.exceptions import RequestDurationException
+
+
+class AsyncRequestType(Enum):
+    MODIFIER = 'modifier'
+    ACCESSOR = 'accessor'
+
+
+class AsyncRequest(object):
+    def __init__(self, kind, library, fun, *args, **kwargs):
+        self.id = uuid.uuid4()
+
+        # Request library call spec
+        self.fun = fun
+        self.args = args
+        self.kwargs = kwargs
+
+        # Request meta
+        self.kind = kind
+        self.library = library
+        self.symbol = kwargs.get('symbol')
+
+        # Request's state
+        self.future = None
+        self.data = None
+        self.exception = None
+        self.is_running = False
+        self.is_completed = False
+
+        # Timekeeping
+        self.start_time = None
+        self.end_time = None
+        self.create_time = time.time()
+
+        self.mongo_retry = bool(kwargs.get('mongo_retry'))
+
+    @property
+    def execution_duration(self):
+        if None in (self.start_time, self.end_time):
+            raise RequestDurationException("{} can't provide an execution_duration ({}).".format(
+                self, (self.start_time, self.end_time)))
+        return self.end_time - self.start_time
+
+    @property
+    def schedule_delay(self):
+        if None in (self.start_time, self.create_time):
+            raise RequestDurationException("{} can't provide an schedule_delay ({}).".format(
+                self, (self.start_time, self.create_time)))
+        return self.start_time - self.create_time
+
+    @property
+    def total_time(self):
+        if None in (self.end_time, self.create_time):
+            raise RequestDurationException("{} can't provide an total_time ({}).".format(
```

Comment:
Looks like it will double parenthethis this

>>> "a({})b".format((1, 2))
'a((1, 2))b'

and above

Reply:
indeed :-)

Reply:
Do you want that?